indexer/file_location_cache.go

```diff
@@ -0,0 +1,159 @@
+	"time"
+
+	"github.com/0glabs/0g-storage-client/common/shard"
+	"github.com/0glabs/0g-storage-client/node"
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const defaultFindFileCooldown = time.Minute * 60
+const defaultDiscoveredURLRetryInterval = time.Minute * 10
+const defaultSuccessCallLifetime = time.Minute * 10
+
+type FileLocationCacheConfig struct {
+	CacheSize     int
+	Expiry        time.Duration
+	DiscoveryNode string
+}
+
+type successCall struct {
+	node *shard.ShardedNode
+	ts   time.Time
+}
+
+type FileLocationCache struct {
+	cache             *expirable.LRU[uint64, []*shard.ShardedNode]
+	latestFindFile    sync.Map // tx seq -> time.Time
+	latestFailedCall  sync.Map // url -> time.Time
+	latestSuccessCall sync.Map // url -> successCall
+	discoverNode      *node.AdminClient
+}
+
+var defaultFileLocationCache FileLocationCache
+
+func InitFileLocationCache(config FileLocationCacheConfig) (closable func(), err error) {
+	if len(config.DiscoveryNode) > 0 {
+		if defaultFileLocationCache.discoverNode, err = node.NewAdminClient(config.DiscoveryNode, defaultZgsClientOpt); err != nil {
+			return nil, errors.WithMessage(err, "Failed to create admin client to discover peers")
+		}
+	}
+	defaultFileLocationCache.cache = expirable.NewLRU[uint64, []*shard.ShardedNode](config.CacheSize, nil, config.Expiry)
+	return defaultFileLocationCache.close, nil
+}
+
+func (c *FileLocationCache) close() {
+	if c.discoverNode != nil {
+		c.discoverNode.Close()
+	}
+}
+
+func (c *FileLocationCache) GetFileLocations(ctx context.Context, txSeq uint64) ([]*shard.ShardedNode, error) {
+	if nodes, ok := c.cache.Get(txSeq); ok {
+		if _, covered := shard.Select(nodes, 1); covered {
+			return nodes, nil
+		}
+	}
+	var nodes []*shard.ShardedNode
+	// fetch from trusted
+	selected := make(map[string]struct{})
+	trusted := defaultNodeManager.TrustedClients()
+	for _, v := range trusted {
+		start := time.Now()
+		fileInfo, err := v.GetFileInfoByTxSeq(ctx, txSeq)
+		if err != nil || !fileInfo.Finalized {
+			continue
+		}
+		config, err := v.GetShardConfig(context.Background())
+		if err != nil || !config.IsValid() {
+			continue
+		}
+		nodes = append(nodes, &shard.ShardedNode{
+			URL:     v.URL(),
+			Config:  config,
+			Latency: time.Since(start).Milliseconds(),
+		})
+		selected[v.URL()] = struct{}{}
+	}
+	logrus.Debugf("find file #%v from trusted nodes, got %v nodes holding the file", txSeq, len(nodes))
+	if _, covered := shard.Select(nodes, 1); covered {
+		return nodes, nil
+	}
+	// trusted nodes do not hold all shards of the file, try to find file
+	if c.discoverNode != nil {
+		locations, err := c.discoverNode.GetFileLocation(ctx, txSeq, false)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Debugf("find file #%v from location cache, got %v nodes holding the file", txSeq, len(locations))
+		for _, location := range locations {
+			url := fmt.Sprintf("http://%v:5678", location.Ip)
```

Comment:
trusted ips may not have this default port I guess.

Reply:
Here are discovered nodes but not trusted, these nodes are discovered by p2p protocol so we just try if they open the default RPC port